tracee-rules/input.go

```diff
@@ -4,70 +4,174 @@ import (
 	"bufio"
 	"encoding/gob"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io"
 	"log"
 	"os"
-	"time"
+	"strings"
 
 	"github.com/aquasecurity/tracee/tracee-rules/types"
 	tracee "github.com/aquasecurity/tracee/tracee/external"
 )
 
-func setupStdinSource(inputSource string) (chan types.Event, error) {
-	res := make(chan types.Event)
-	scanner := bufio.NewScanner(os.Stdin)
-	go func() {
-		for scanner.Scan() {
-			event := scanner.Bytes()
-			switch inputSource {
-			case "tracee":
-				var e tracee.Event
-				err := json.Unmarshal(event, &e)
-				if err != nil {
-					log.Printf("invalid json in %s: %v", string(event), err)
-				}
-				res <- types.Event(e)
-			}
-		}
-		close(res)
-	}()
-	return res, nil
+var errHelp = errors.New("user has requested help text")
+
+type inputFormat uint8
+
+const (
+	invalidInputFormat inputFormat = iota
+	jsonInputFormat
+	gobInputFormat
+)
+
+type traceeInputOptions struct {
+	inputFile   *os.File
+	inputFormat inputFormat
 }
 
-func setupTraceeSource(traceeFilePath string) (chan types.Event, error) {
-	_, err := os.Stat(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid Tracee input file: %s", traceeFilePath)
+func setupTraceeInputSource(opts *traceeInputOptions) (chan types.Event, error) {
+
+	if opts.inputFormat == jsonInputFormat {
+		return setupTraceeJSONInputSource(opts)
 	}
-	f, err := os.Open(traceeFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid file: %s", traceeFilePath)
+
+	if opts.inputFormat == gobInputFormat {
+		return setupTraceeGobInputSource(opts)
 	}
-	dec := gob.NewDecoder(f)
+
+	return nil, errors.New("could not set up input source")
+}
+
+func setupTraceeGobInputSource(opts *traceeInputOptions) (chan types.Event, error) {
```

Comment:
This is mostly a question for myself but why does this simple test program not work?

```
func Test_setupTraceeGobInputSource(t *testing.T) {
	f, err := ioutil.TempFile("", "Test_setupTraceeGobInputSource-*")
	require.NoError(t, err)
	defer func() {
		_ = f.Close()
		_ = os.RemoveAll(f.Name())
	}()

	// write to file
	e := gob.NewEncoder(f)
	event := tracee.Event{
		EventName: "foo",
	}
	require.NoError(t, e.Encode(event))
	
	opts := traceeInputOptions{
		inputFile:   f,
		inputFormat: gobInputFormat,
	}

	res, err := setupTraceeGobInputSource(&opts)
	require.NoError(t, err)
	assert.Equal(t, "foo", (<-res).(tracee.Event).EventName)
}
```